java-support/src/main/scala/io/cloudstate/javasupport/impl/controller/AnnotationBasedControllerSupport.scala

```
@@ -0,0 +1,462 @@
+        } else annotation.name()
+
+        val serviceMethod = serviceMethods.getOrElse(name, {
+          throw new RuntimeException(
+            s"Command handler method ${method.getName} for command $name found, but the service has no command by that name."
+          )
+        })
+
+        (method, serviceMethod)
+      }
+      .groupBy(_._2.name)
+      .map {
+        case (commandName, Seq((method, serviceMethod))) => (commandName, method, serviceMethod)
+        case (commandName, many) =>
+          throw new RuntimeException(
+            s"Multiple methods found for handling command of name $commandName: ${many.map(_._1.getName).mkString(", ")}"
+          )
+      }
+
+    val unaryCallHandlers = allCallHandlers.collect {
+      case (commandName, method, serviceMethod)
+          if !serviceMethod.descriptor.isClientStreaming && !serviceMethod.descriptor.isServerStreaming =>
+        commandName -> new UnaryCallInvoker(method, serviceMethod)
+    }.toMap
+
+    val serverStreamedCallHandlers = allCallHandlers.collect {
+      case (commandName, method, serviceMethod)
+          if !serviceMethod.descriptor.isClientStreaming && serviceMethod.descriptor.isServerStreaming =>
+        commandName -> new ServerStreamedCallInvoker(method, serviceMethod)
+    }.toMap
+
+    val clientStreamedCallHandlers = allCallHandlers.collect {
+      case (commandName, method, serviceMethod)
+          if serviceMethod.descriptor.isClientStreaming && !serviceMethod.descriptor.isServerStreaming =>
+        commandName -> new ClientStreamedCallInvoker(method, serviceMethod, mat)
+    }.toMap
+
+    val streamedCallHandlers = allCallHandlers.collect {
+      case (commandName, method, serviceMethod)
+          if serviceMethod.descriptor.isClientStreaming && serviceMethod.descriptor.isServerStreaming =>
+        commandName -> new StreamedCallInvoker(method, serviceMethod, mat)
+    }.toMap
+
+    ReflectionHelper.validateNoBadMethods(
```

just a small change. perhaps it can be done first to fail fast after allMethods are processed.